lib/ansible/plugins/strategy/__init__.py

```diff
@@ -1007,6 +1019,7 @@ def _evaluate_conditional(h):
             if task.when:
                 self._cond_not_supported_warn(meta_action)
             self._inventory.refresh_inventory()
+            self._set_hosts_cache(iterator._play)
```

Comment:
This one's probably also suspect, though in this case I think we want to *remove* hosts but not add new ones. It really depends on how we expect plays to behave when hosts appear/disappear mid-play, which doesn't seem to be perfectly defined.

Reply:
I think this one is correct.  `inventory.refresh_inventory()` fully clears the inventory and re-populates it.  A call to `get_hosts` would have then re-populated based on the new state of the inventory which would have both added new hosts from inventory sources and removed hosts.

Reply:
It might be correct in that it matches what was happening before, but there seemed to be some disagreement about what exactly should be in the `ansible_play_hosts` vars when the inventory changes mid-play. Preserving the pre-Perfy behavior is probably best for now, but it's probably a decision we should make explicitly (and if we were designing it all from scratch, I'd argue that the vars should reflect what's actually happening in the current play, which they sometimes don't today). I tripped over several "unintuitive" dynamic inventory behaviors while testing this stuff (most of which existed previously)- we don't need to correct them here, but I also don't want to forget about them, so if we're not going to fix them, I'll open separate issues for them.

Reply:
Due to the potential of unintended consequences of this PR already, I'd like to limit behavior changes here if possible.